### PR TITLE
Implement abstract operation `GetPrototypeFromConstructor`

### PR DESCRIPTION
**Pull request by jedel1043**
This Pull Request reduces the repetition on constructors when obtaining the default prototype of an object.
Extracted the pattern from #1552, thanks to @Razican :)

It changes the following:

- Implements the abstract operation `GetPrototypeFromConstructor`

- Replaces the pattern:
```Rust
let proto = new_target
        .as_object()
        .and_then(|obj| {
            obj.__get__(&PROTOTYPE.into(), obj.clone().into(), context)
                .map(|o| o.as_object())
                .transpose()
        })
        .transpose()?
        .unwrap_or_else(default_proto);
```
with
```Rust
let proto = get_prototype_from_constructor(
                new_target,
                |intrinsics| intrinsics.object_object().prototype(),
                context,
            )?;
```
This also allows for an easy replacement of  `intrinsics` with the intrinsic objects of a `realm` when we eventually implement multiple realms.

### CODE REVIEW

**Comment by Razican** [comment]
Should this be implemented as part of `Context`?

**Comment by raskad** [comment]
I'm not sure if the `intrinsicDefaultProto` should just be a `FnOnce(&Context) -> JsObject`. It is not probable, but I think this function could be misused with that definition. I think if we implement this, we should honor the `1. Assert: intrinsicDefaultProto is this specification's name of an intrinsic object.` somewhat. Maybe we could use an enum as the argument and do the `Context.standard_objects().some_object().prototype()` in the function?

The spec defines this as an ordinary object internal method, so I think the place is good.

File: boa/src/builtins/boolean/mod.rs [hunk 1]
```diff
@@ -66,15 +66,11 @@ impl Boolean {
...
-            .and_then(|obj| {
-                obj.__get__(&PROTOTYPE.into(), obj.clone().into(), context)
-                    .map(|o| o.as_object())
-                    .transpose()
-            })
-            .transpose()?
-            .unwrap_or_else(|| context.standard_objects().object_object().prototype());
+        let prototype = get_prototype_from_constructor(
+            new_target,
+            |intrinsics| intrinsics.object_object().prototype(),
```
Comment (raskad) [inline boa/src/builtins/boolean/mod.rs:71]: Should this be `boolean_object()`?

File: boa/src/builtins/function/mod.rs [hunk 2]
```diff
@@ -302,15 +303,11 @@ impl BuiltInFunctionObject {
...
-            .and_then(|obj| {
-                obj.__get__(&PROTOTYPE.into(), obj.clone().into(), context)
-                    .map(|o| o.as_object())
-                    .transpose()
-            })
-            .transpose()?
-            .unwrap_or_else(|| context.standard_objects().object_object().prototype());
+        let prototype = get_prototype_from_constructor(
+            new_target,
+            |intrinsics| intrinsics.object_object().prototype(),
```
Comment (raskad) [inline boa/src/builtins/function/mod.rs:308]: Should this be `function_object()`?

File: boa/src/builtins/number/mod.rs [hunk 3]
```diff
@@ -166,15 +166,11 @@ impl Number {
...
-            .and_then(|obj| {
-                obj.__get__(&PROTOTYPE.into(), obj.clone().into(), context)
-                    .map(|o| o.as_object())
-                    .transpose()
-            })
-            .transpose()?
-            .unwrap_or_else(|| context.standard_objects().object_object().prototype());
+        let prototype = get_prototype_from_constructor(
+            new_target,
+            |intrinsics| intrinsics.object_object().prototype(),
```
Comment (raskad) [inline boa/src/builtins/number/mod.rs:171]: Should this be `number_object()`?

File: boa/src/builtins/string/mod.rs [hunk 4]
```diff
@@ -179,16 +180,11 @@ impl String {
...
-            .and_then(|obj| {
-                obj.__get__(&PROTOTYPE.into(), obj.clone().into(), context)
-                    .map(|o| o.as_object())
-                    .transpose()
-            })
-            .transpose()?
-            .unwrap_or_else(|| context.standard_objects().object_object().prototype());
+        let prototype = get_prototype_from_constructor(
+            new_target,
+            |intrinsics| intrinsics.object_object().prototype(),
```
Comment (raskad) [inline boa/src/builtins/string/mod.rs:185]: Should this be `string_object()`?

**Review by raskad — COMMENTED** [review]
Looks good to me.

I commented on a few places where we use `object_object()` while a specific `StandardConstructor` exists. I'm curious if changing those would result in actual changes.

**Comment by jedel1043** [comment]
> Looks good to me.
> 
> I commented on a few places where we use `object_object()` while a specific `StandardConstructor` exists. I'm curious if changing those would result in actual changes.

Let's see if this works

**Review by raskad — APPROVED** [review]
Nice work!

**Comment by jedel1043** [comment]
> Nice work!

Good catch!

**Comment by Razican** [comment]
I think it would be easier to implement it by using `FnOnce(&StandardObjects) -> &StandardConstructor`. I did this in #1552 in my latest changes, but it would allow passing `StandardObjects::typed_array_object` and things like that, so directly the function.

**Comment by jedel1043** [comment]
> I think it would be easier to implement it by using `FnOnce(&StandardObjects) -> &StandardConstructor`. I did this in #1552 in my latest changes, but it would allow passing `StandardObjects::typed_array_object` and things like that, so directly the function.

Yep, this is the best option.

**Review by Razican — APPROVED** [review]
Now it looks great! :) thanks!

**Pull request merged**